pkg/pillar/cmd/zedagent/handlecertconfig.go

```diff
@@ -199,19 +325,105 @@ func sendCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
 	_, _, rtf, err := zedcloud.SendOnAllIntf(&zedcloudCtx, attestURL,
 		size, buf, iteration, bailOnHTTPErr)
 	if err != nil {
+		// Hopefully next timeout will be more successful
 		if rtf == types.SenderStatusRemTempFail {
-			log.Errorf("sendCertsProtobuf remoteTemporaryFailure: %s",
+			log.Errorf("sendEveNodeCertsProtobuf remoteTemporaryFailure: %s",
 				err)
 		} else {
-			log.Errorf("sendCertsProtobuf failed: %s", err)
+			log.Errorf("sendEveNodeCertsProtobuf failed: %s", err)
 		}
-		// Try sending later
-		// The buf might have been consumed
-		buf := bytes.NewBuffer(data)
-		if buf == nil {
-			log.Fatal("malloc error")
-		}
-		zedcloud.SetDeferred("attest:"+deviceUUID, buf, size, attestURL,
+		zedcloud.SetDeferred("attest:"+reqID, buf, size, attestURL,
 			zedcloudCtx, true)
 	}
 }
+
+// initialize cipher pubsub trigger handlers and channels
+func cipherModuleInitialize(ctx *zedagentContext, ps *pubsub.PubSub) {
+
+	// create the trigger channels
+	ctx.cipherCtx.triggerEveNodeCerts = make(chan struct{}, 1)
+	ctx.cipherCtx.triggerControllerCerts = make(chan struct{}, 1)
+}
+
+// start the task threads
+func cipherModuleStart(ctx *zedagentContext) {
+	if !zedcloud.UseV2API() {
+		log.Infof("V2 APIs are still not enabled")
+		// we will run the tasks for watchdog
+	}
+	// start the eve node certificate push task
+	go eveNodeCertsTask(ctx, ctx.cipherCtx.triggerEveNodeCerts)
+
+	// start the controller certificate fetch task
+	go controllerCertsTask(ctx, ctx.cipherCtx.triggerControllerCerts)
+}
+
+// Controller certificate, check whether there is a Sha mismatch
+// to trigger the post request
+func handleControllerCertsSha(ctx *zedagentContext,
+	config *zconfig.EdgeDevConfig) {
+	sumHash := hex.EncodeToString(ctx.cipherCtx.cfgControllerCertHash)
+	certHash := config.GetControllercertConfighash()
+	if sumHash != certHash {
+		triggerControllerCertEvent(ctx)
+	}
+}
+
+// Eve node certificate, check whether there is a Sha mismatch
+// to trigger the post request
+func handleEveNodeCertsSha(ctx *zedagentContext,
+	config *zconfig.EdgeDevConfig) {
+	sumHash := hex.EncodeToString(ctx.cipherCtx.eveNodeCertHash)
+	certHash := config.GetNodecertConfighash()
```

Comment:
Ditto, since the controller doesn't yet implement this.

Reply:
Actually it is better to simplify this by removing the ctx.cipherCtx.eveNodeCertHash. pubsub will only call the handleModify when there is a change which means you need to trigger sending a message.

Reply:
sure